### PR TITLE
http_filter: Add gRPC JSON reverse transcoder filter

### PR DESCRIPTION
**Pull request by numanelahi**
**Commit Message:** http_filter: Add gRPC JSON reverse transcoder filter. 
### Additional Description
The gRPC JSON reverse transcoder acts as a bridge between gRPC client and an HTTP/JSON server, converting the gRPC request into HTTP/JSON or the HTTP backend and the HTTP/JSON response back to gRPC for the gRPC client.

Thanks @yanavlasov for reviewing the google's internal design doc for gRPC JSON reverse transcoder filter.

**Risk Level:** Low
**Testing:** Added unit and integrations tests.
**Docs Changes:** Nil
**Release Notes:** Nil
**Platform Specific Features:** Nil

### CODE REVIEW

File: docs/root/configuration/http/http_filters/_include/grpc-json-reverse-transcoder-filter.yaml
Comment (phlax) [inline docs/root/configuration/http/http_filters/_include/grpc-json-reverse-transcoder-filter.yaml:1]: is admin required for this example?

File: bazel/repository_locations.bzl
Comment (phlax) [inline bazel/repository_locations.bzl:988]: deps ci is saying:

```console
Extension envoy.filters.http.grpc_json_reverse_transcoder depends on grpc_httpjson_transcoding but grpc_httpjson_transcoding does not list envoy.filters.http.grpc_json_reverse_transcoder in its allowlist
```

File: test/extensions/filters/http/grpc_json_reverse_transcoder/utils_test.cc
Comment (phlax) [inline test/extensions/filters/http/grpc_json_reverse_transcoder/utils_test.cc:126]: struggling to understand why but this is triggering a spellcheck failure

File: test/extensions/filters/http/grpc_json_reverse_transcoder/filter_test.cc
Comment (phlax) [inline test/extensions/filters/http/grpc_json_reverse_transcoder/filter_test.cc:224]:
```suggestion
// Test request transcoding where request type is `google.api.HttpBody`
```

not certain, but i think marking as literal will fix _this_ spellcheck failure

File: docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst
Comment (phlax) [inline docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst:19]: this anchor is duplicating an existing one

File: docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst
Comment (phlax) [inline docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst:2]: this file isnt currently included in any (rst) toctree

**Review by phlax — COMMENTED** [review]
thanks for working on this @numanelahi

if you can fix the docs issues i can review the docs part

also coverage is saying:

```console
Code coverage for source/extensions/filters/http/grpc_json_reverse_transcoder is lower than limit of 96.6 (94.8)
```

/coverage

File: test/extensions/filters/http/grpc_json_reverse_transcoder/filter_test.cc
Comment (numanelahi) [reply test/extensions/filters/http/grpc_json_reverse_transcoder/filter_test.cc]: Fixed

File: test/extensions/filters/http/grpc_json_reverse_transcoder/utils_test.cc
Comment (numanelahi) [reply test/extensions/filters/http/grpc_json_reverse_transcoder/utils_test.cc]: It's because of the `audobook`; changed it to `audio-book` to fix this.

File: bazel/repository_locations.bzl
Comment (numanelahi) [reply bazel/repository_locations.bzl:988]: Fixed.

File: docs/root/configuration/http/http_filters/_include/grpc-json-reverse-transcoder-filter.yaml
Comment (numanelahi) [reply docs/root/configuration/http/http_filters/_include/grpc-json-reverse-transcoder-filter.yaml]: Thanks for pointing out; it's not required, so removed it.

File: docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst
Comment (numanelahi) [reply docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst]: Fixed

**Comment by numanelahi** [comment]
> thanks for working on this @numanelahi
> 
> if you can fix the docs issues i can review the docs part
> 
> also coverage is saying:
> 
> ```
> Code coverage for source/extensions/filters/http/grpc_json_reverse_transcoder is lower than limit of 96.6 (94.8)
> ```
> 
> /coverage

@phlax  Currently, the filters code coverage is 95.7%. It's hard to take it beyond this. Is there a way to get an exception for this or I've to increase it beyond 96.6%?

**Comment by phlax** [comment]
you can add an exception in `test/per_file_coverage.sh` but it will need sign off from coverage sheps

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (abeyad) [inline api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:22]: nit: s/between gRPC/between a gRPC

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (abeyad) [inline api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:27]: usage of `oneof` is discouraged: https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md?plain=1#L63-L91.

I assume this was borrowed from how https://github.com/envoyproxy/envoy/blob/c1798d7f9cd36a27cf59ea390bdbe702aea0be3a/api/envoy/extensions/filters/http/grpc_json_transcoder/v3/transcoder.proto#L113 does it, but I'm thinking it would be better to define two separate fields and an order of precedence to conform to the style guide recommendations.

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (abeyad) [inline api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:21]: it might make sense to add a note that this is the reverse of https://github.com/envoyproxy/envoy/blob/c1798d7f9cd36a27cf59ea390bdbe702aea0be3a/docs/root/configuration/http/http_filters/grpc_json_transcoder_filter.rst which allows JSON client -> gRPC service.  That way the "reverse" in the name would make more sense.

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (abeyad) [inline api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:26]: I know "reverse" here makes sense since there is already a `GrpcJsonTranscoder`, but I'm wondering if a better name would be `GrpcToJsonTranscoder`, making the direction clear.

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (abeyad) [inline api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:47]: under what situations would it make sense to not just change the max stream buffer size instead?

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (abeyad) [inline api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:55]: same question as above, when would it be useful to not just change the stream buffer size?

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (abeyad) [inline api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:57]: which API version is this referring to?
also, nit: s/api/API

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (numanelahi) [reply api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:26]: `GrpcToJsonTranscoder` may make it sound bit confusing, as it doesn't spell that different from `GrpcJsonTranscoder` and feels like both the filters are doing the same thing. That's why I think "reverse" makes more sense.

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (numanelahi) [reply api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:47]: I not sure if I understand what you're trying to ask, but `max_request_body_size` is the max buffer stream size. If this is not set the filter will use the default stream buffer limit.

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (numanelahi) [reply api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:55]: Same as above.

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (numanelahi) [reply api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:21]: Done

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (numanelahi) [reply api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:57]: The is referring to the service's api version, in case of JSON to gRPC transcoding it's part of the path itself, {api_version}/resource/{resource_id}. But, in case of gRPC to JSON transcoding we're supposed to construct this path from the gRPC request message with usually doesn't contain the api_version so this has to come from the header. So, the config field is the name of the header that we're supposed to read to get the header from.

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (numanelahi) [reply api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto]: Done

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (abeyad) [reply api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:28]: > GrpcToJsonTranscoder may make it sound bit confusing, as it doesn't spell that different from GrpcJsonTranscoder 

That's a fair point.

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (abeyad) [reply api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:47]: > max_request_body_size is the max buffer stream size

should this be renamed then to max_buffer_stream_size ?  what happens if the request body sent by the client is larger than this value?

Comment (numanelahi) [reply api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:47]: The filter buffers the request/response before forwarding it upstream/downstream. If the request/response exceeds the stream buffer limit then the filter stop the iteration and buffering and returns the local response with ResourceExhausted/Internal status. 

Here the max_request/response_body_size is the upper limit on the request/response sizes, which for filter means we don't wanna buffer the data more than the max allowed request/response sizes, that's why we're setting the buffer limits to the values of these fields.

Comment (abeyad) [reply api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:47]: > The filter buffers the request/response before forwarding it upstream/downstream. If the request/response exceeds the stream buffer limit then the filter stop the iteration and buffering and returns the local response with ResourceExhausted/Internal status.

This is good info, thanks!  I suggest putting it in the comments for the field.  In that case, `max_request_body_size` makes sense for a name.

Comment (numanelahi) [reply api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:48]: Updated the comments on both the fields.

**Review by abeyad — APPROVED** [review]
/lgtm api

thanks @numanelahi !

**Comment by numanelahi** [comment]
@phlax Can you please please finish your review, or you're waiting for the approval from coverage sheps.

**Review by RyanTheOptimist — COMMENTED** [review]
coverage lgtm

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (phlax) [inline api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:21]:
```suggestion
// ``GrpcJsonReverseTranscoder`` is the filter configuration for the gRPC JSON
```

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (phlax) [inline api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto:32]:
```suggestion
  // If set, takes precedence over the ``descriptor_binary`` field.
```

File: docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst
Comment (phlax) [inline docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst:27]:
```suggestion
googleapis repository from GitHub before running ``protoc``, as you'll need annotations.proto
```

File: docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst
Comment (phlax) [inline docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst:26]: This should use the googleapis that are provided by Envoy's bazel setup - same with protoc - its provided by Envoys build system

File: docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst
Comment (phlax) [inline docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst:35]:
```suggestion
Then run ``protoc`` to generate the descriptor set. For example using the test
```

File: docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst
Comment (phlax) [inline docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst:52]: This includes the entire file - what would be better is to just include the lines that are relevant, and optionally highlight any that are discussed in the text

File: docs/root/configuration/http/http_filters/_include/grpc-json-reverse-transcoder-filter.yaml
Comment (phlax) [inline docs/root/configuration/http/http_filters/_include/grpc-json-reverse-transcoder-filter.yaml:7]: i would suggest using port 10000 for this as that is the common default - it would also make it less confusing in relation to the upstream port

**Review by phlax — COMMENTED** [review]
Docs review bel

wrt protoc/googleapis - im not sure how best to document that as a bazel build but i think it would be better to re/use the existing dependencies rather than have users download versions that may/not work

this also needs a changelog - and will require a final code signoff (from someone other than me)

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (numanelahi) [reply api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto]: Done

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (numanelahi) [reply api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto]: Done

File: docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst
Comment (numanelahi) [reply docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst]: Done

File: docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst
Comment (numanelahi) [reply docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst:26]: As far as I know `protoc` looks for the actual proto files and not the bazel targets, so how can that be used with `protoc`, CMIIAW.

File: docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst
Comment (numanelahi) [reply docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst]: Done

File: docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst
Comment (numanelahi) [reply docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst:52]: Done

File: docs/root/configuration/http/http_filters/_include/grpc-json-reverse-transcoder-filter.yaml
Comment (numanelahi) [reply docs/root/configuration/http/http_filters/_include/grpc-json-reverse-transcoder-filter.yaml]: Done

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (numanelahi) [reply api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto]: Done

File: api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto
Comment (numanelahi) [reply api/envoy/extensions/filters/http/grpc_json_reverse_transcoder/v3/transcoder.proto]: Done

File: docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst
Comment (numanelahi) [reply docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst]: Done

File: docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst
Comment (numanelahi) [reply docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst]: Done

**Comment by numanelahi** [comment]
> this also needs a changelog 

Added the change log

> - and will require a final code signoff (from someone other than me)

Can you suggest who should be added for that, and I don't have the authorization to add the reviewers.

File: docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst
Comment (phlax) [reply docs/root/configuration/http/http_filters/grpc_json_reverse_transcoder_filter.rst:36]: im struggling to find a good (standalone) example

a basic example usage of the envoy configured `protoc`:

```console
$ bazel run @com_google_protobuf//:protoc -- --version
```
this is an example of `run` usage - to use it together with the envoy supplied googleapis dep you kinda need to figure out a `build` pattern

the googleapis are available from `@com_google_googleapis` - if you grep the envoy code for that string you will see the various ways these are used

the complexity is that i guess you want to combine `protoc` + `googleapis` + user code - in that case you would need to add a `build` target with a configurable input

trying to think if there are any existing patterns to copy - probably there is, but cant think of any right now

**Comment by phlax** [comment]
/assign @alyssawilk 

... as `grpc_json_transcoder` CODEOWNER

**Comment by alyssawilk** [comment]
I'm going to throw it over to Yan as he's the listed codeowner. If he didn't sign on to support this should land in contrib I think?

File: source/extensions/extensions_build_config.bzl
Comment (yanavlasov) [inline source/extensions/extensions_build_config.bzl:1]: It looks like this file god massively reformatted. Can you revert the unnecessary changes, please?

File: source/extensions/filters/http/grpc_json_reverse_transcoder/filter.h
Comment (yanavlasov) [inline source/extensions/filters/http/grpc_json_reverse_transcoder/filter.h:111]: You may want to expose counters for these buffers in the future, given that you have separate control over their sizes. Otherwise you limit visibility into memory usage.

Also this pattern will prevent Envoys overload manager from resetting most expensive streams under memory pressure, since it will not have visibility into these buffers. There is nothing you can do at this point about it, just pointing out why making memory usage of these buffers visible is important.

File: source/extensions/filters/http/grpc_json_reverse_transcoder/filter_config.h
Comment (yanavlasov) [inline source/extensions/filters/http/grpc_json_reverse_transcoder/filter_config.h:86]: I strongly recommend making all members of the config object const, since it is shared across multiple threads. This will help to be ensure that it is immutable and hence thread safe.

**Review by yanavlasov — COMMENTED** [review]
/wait

File: source/extensions/filters/http/grpc_json_reverse_transcoder/filter_config.h
Comment (numanelahi) [reply source/extensions/filters/http/grpc_json_reverse_transcoder/filter_config.h:86]: Changed the type of `config_` in filter class from `std::shared_ptr<GrpcJsonReverseTranscoderConfig>` to  `std::shared_ptr<cosnt GrpcJsonReverseTranscoderConfig>`, this should make it immutable.

File: source/extensions/filters/http/grpc_json_reverse_transcoder/filter.h
Comment (numanelahi) [reply source/extensions/filters/http/grpc_json_reverse_transcoder/filter.h:111]: Acknowledged!

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
